Use a template to produce the default help message

mkantor opened the pull request:
This makes some changes to the template system:

- Template tags for optional items (like `{author}`) now expand to nothing when the value is unset instead of a default string (like `"unknown author"`).
- More template tags now emit line-wrapped output to match the behavior of `write_default_help`.
- Items with long variants now expand to the appropriate thing for `-h` vs `--help`.
- The now-obsolete `{long-about}` tag has been removed.
- A few new tags have been added.

These are externally-visible changes, but if this makes it into 3.0 that's probably reasonable?

Note that line-wrapping can have some odd edge cases since it does not account for preceding/trailing characters on the same line as the tag. This is already the case in `master`, but will affect some additional tags with this changeset. See #2065 for details.

An alternative approach would be to make the template language itself fancier, adding explicit mechanisms to control wrapping and/or conditional output. But since the template system is "by design very simple" this could be undesirable.

Closes #2002.

pksunkara commented on src/build/app/mod.rs:
````diff
@@ -754,9 +758,6 @@ impl<'help> App<'help> {
     ///     .help_template("{bin} ({version}) - {usage}")
     /// # ;
     /// ```
-    /// [`App::about`]: ./struct.App.html#method.about
````
I think these should exist and the links to them in the above description such that people will have an idea where there are coming from.

mkantor replied:
Sure. I removed them because there's missing nuance; e.g. sometimes it's `long_about`. Do you think I should mention that or just keep it simple?

pksunkara replied:
Let's say about or long_about and link them both. No need to explain more.

mkantor replied:
Made some edits in 2c91800.

pksunkara commented on src/build/app/mod.rs:
```diff
@@ -724,26 +724,30 @@ impl<'help> App<'help> {
+    ///   * `{author}`              - Author information.
+    ///   * `{author-with-newline}` - Author followed by `\n`.
+    ///   * `{about}`               - Description of the program.
+    ///   * `{about-with-newline}`  - Description followed by `\n`.
```
I don't think this and the `author-with-newline` should exist. The templates can have additional `\n` with the normal `about` and `author`.

mkantor replied:
Trouble is, the `\n` has to be conditional. If about/author is not set then there is no output, otherwise it's the value plus a newline.

Originally I had multiple templates and chose between them, but quickly realized there's a combinatorial explosion with things like before/after help, etc. And building up a template dynamically seemed like it defeated the point.

The "right" way to do this is to extend the template language to support conditionals (e.g. `{if author}{author}\n{/if}`), but keeping the language simple is a great goal so I didn't go there.

Thoughts?

pksunkara replied:
I suspected something like that being the reason. Looks like we will have to be satisfied with this for now.

pksunkara commented on src/build/app/mod.rs:
```diff
@@ -724,26 +724,30 @@ impl<'help> App<'help> {
+    ///   * `{after-help}`          - Info to be displayed after the help message.
+    ///   * `{after-help-padded}`   - After help message with additional spacing.
+    ///   * `{before-help}`         - Info to be displayed before the help message.
+    ///   * `{before-help-padded}`  - Before help message with additional spacing.
```
I don't think this and `after-help-padded` should exist. The normal ones should always be padded.

mkantor replied:
Done in 059503e.

pksunkara commented on src/output/help.rs:
```diff
@@ -1083,16 +977,41 @@ impl<'help, 'app, 'parser, 'writer> Help<'help, 'app, 'parser, 'writer> {
+                        self.parser.app.about
+                    };
+                    if let Some(output) = about {
+                        self.none(output)?;
```
`wrap_help` should be used.

mkantor replied:
See 059503e.

pksunkara commented on src/output/help.rs:
```diff
@@ -1083,16 +977,41 @@ impl<'help, 'app, 'parser, 'writer> Help<'help, 'app, 'parser, 'writer> {
                 b"author" => {
-                    self.none(self.parser.app.author.unwrap_or("unknown author"))?;
+                    if let Some(output) = self.parser.app.author {
+                        self.none(output)?;
```
`wrap_help` should be used.

mkantor replied:
See 059503e.

pksunkara commented on src/output/help.rs:
```diff
@@ -1124,10 +1043,29 @@ impl<'help, 'app, 'parser, 'writer> Help<'help, 'app, 'parser, 'writer> {
                 b"before-help" => {
-                    self.none(self.parser.app.before_help.unwrap_or("unknown before-help"))?;
+                    let before_help = if self.use_long {
+                        self.parser.app.before_long_help
```
`.or(self.parser.app.before_help)`

mkantor replied:
Oops, nice catch. Fixed in 4545a47.

pksunkara commented on src/build/app/mod.rs:
```diff
@@ -724,26 +724,29 @@ impl<'help> App<'help> {
+    ///   * `{author}`              - Author information.
+    ///   * `{author-with-newline}` - Author followed by `\n`.
+    ///   * `{about}`               - General description (from [`App::about`] or
+    ///                               [`App::long_about`]).
```
You don't need a line break here.

mkantor replied:
What column width should be wrapped at? I thought it was 100 characters based on what the comment looked like before.

pksunkara replied:
Oh sorry. My bad then.

pksunkara commented on src/output/help.rs:
```diff
@@ -1124,10 +1043,29 @@ impl<'help, 'app, 'parser, 'writer> Help<'help, 'app, 'parser, 'writer> {
+                        self.parser.app.after_help
+                    };
+                    if let Some(output) = after_help {
+                        self.none("\n\n")?;
```
Can you add a small test that generates help of a program which has after_help but no args at all? I think you will see that there will be 4 empty lines because [this](https://github.com/clap-rs/clap/pull/2067/files#diff-750408d27c2eebbe9c64eccbc80aac19L915-L917) condition is not used.

mkantor replied:
Good eye! That scenario does indeed produce 4 empty lines, but only when the default args for `--help` and `--version` are also disabled.

I can fix it by [using a different template when there are no args](https://github.com/mkantor/clap/compare/default-template...WIP/default-template-different-no-arg-template).

Alternatively, I could:

1. [Check for args while expanding `{after-help}`, don't emit `\n\n` when there aren't any](https://github.com/mkantor/clap/compare/default-template...WIP/default-template-magic-padding-for-after-help).

   Downside: this bakes assumptions about the overall template into `{after-help}` (namely that it will always come right after `{all-args}`). That's the case for the default template, but might not be for user-defined templates.

1. Do nothing; leave the code as-is.

   If this configuration (no user-defined args, `--help` and `--version`  disabled, and `after_help` set to something) is highly unusual, maybe it's okay if it looks a little weird.

Let me know what you prefer.

pksunkara replied:
Using a different template sounds like the correct solution.

mkantor replied:
Done: 85d3daa.

mkantor commented:
@pksunkara Thanks for all the super quick & helpful feedback so far! I've got some other obligations to attend to but will continue addressing your comments later. Cheers till then!
